model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/UserSessionQueryConditionalRemover.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.sessions.infinispan.changes.remote.remover.query;
+
+import java.util.Map;
+import java.util.Objects;
+
+import org.keycloak.models.sessions.infinispan.changes.remote.remover.ConditionalRemover;
+import org.keycloak.models.sessions.infinispan.entities.RemoteUserSessionEntity;
+import org.keycloak.models.sessions.infinispan.query.UserSessionQueries;
+
+/**
+ * A {@link ConditionalRemover} implementation to remove {@link RemoteUserSessionEntity} based on some filters over its
+ * state.
+ * <p>
+ * This implementation uses Infinispan Ickle Queries to perform the removal operation. Indexing is not required.
```

Comment:
When we use Ickle queries, how to these statements perform with an increasing number of user sessions and client sessions? On which queries do we use indexes, on which queries will we do a full scan of all entries in memory or in store (I know we don't use a persistent store in Infinispan yet)?

Reply:
I annotated all fields used in these queries to be indexed. If indexes are enabled in the external Infinispan, it will perform indexed queries. If not, it will do full scans. Hybrid queries should not happen at all. 

For the performance, we have to test it for KC 27 (volatile sessions won't be supported in KC 26, so this code will never be invoked). 
In the support case I worked on a couple of months ago, they were using indexed queries. With a total of 1'800'000 entries and a 6 nodes cluster (~300'000 entries per node), the indexed queries took an average of 7ms on each node. Non-indexed queries (full scans) were taking closer to 100ms if my memory is not failing me right now.

Reply:
Thank you for these numbers. I read them that we would see a noticeable delay once we cross 100_000 entries per node. 

Given those numbers, I suggest to create a follow-up issue about indexing, and then merge the PR as is.